File: src/leetcode/valid_sudoku.py
from typing import List
from collections import defaultdict


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # keys are row number, column number, or square number
        # values are sets of numbers in the row, column, or square
        row_dict = defaultdict(set)
        col_dict = defaultdict(set)
        square_dict = defaultdict(set)
        # divide the board into 9 squares
        # the key will be a tuple e.g., (0, 0) for the top-left square

        # 0, 0 | 0, 1 | 0, 2
        # ------------------
        # 1, 0 | 1, 1 | 1, 2
        # ------------------
        # 2, 0 | 2, 1 | 2, 2

        # Remark:
        # using absolute row/column indices (0-8) wouldn't work as well
        # for tracking 3x3 squares
        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num == ".":
                    continue
                elif (
                    num in row_dict[i]
                    or num in col_dict[j]
                    or num in square_dict[(i // 3, j // 3)]  # square number
                ):
                    return False
                row_dict[i].add(num)
                col_dict[j].add(num)
                square_dict[(i // 3, j // 3)].add(num)

        return True
    
# Example usage
if __name__ == "__main__":
    solution = Solution()
    print(solution.isValidSudoku(board = 
        [["1","2",".",".","3",".",".",".","."],
        ["4",".",".","5",".",".",".",".","."],
        [".","9","8",".",".",".",".",".","3"],
        ["5",".",".",".","6",".",".",".","4"],
        [".",".",".","8",".","3",".",".","5"],
        ["7",".",".",".","2",".",".",".","6"],
        [".",".",".",".",".",".","2",".","."],
        [".",".",".","4","1","9",".",".","8"],
        [".",".",".",".","8",".",".","7","9"]]
        ))  # True

 

File: src/leetcode/julia/stack/ValidSudoku.jl
function isValidSudoku(board::Vector{Vector{Char}})::Bool
    row_dict = Dict{Int, Set{Char}}()
    col_dict = Dict{Int, Set{Char}}()
    box_dict = Dict{Tuple{Int,Int}, Set{Char}}()

    # (1,1) | (1,2) | (1,3)    # i÷3+1, j÷3+1 coordinates
    # -------|-------|-------
    # (2,1) | (2,2) | (2,3)
    # -------|-------|-------
    # (3,1) | (3,2) | (3,3)

    for i in 1:9
        for j in 1:9
            num = board[i][j]
            num == '.' && continue

            if num in get!(row_dict, i, Set{Char}()) ||
               num in get!(col_dict, j, Set{Char}()) ||
               num in get!(box_dict, (i÷3, j÷3), Set{Char}())
                return false
            end
            
            push!(row_dict[i], num)
            push!(col_dict[j], num)
            push!(box_dict[(i÷3, j÷3)], num)
            end
        end
        return true
    end

# Example usage
if abspath(PROGRAM_FILE) == @__FILE__
    board = 
    [['1','2','.','.','3','.','.','.','.'],
    ['4','.','.','5','.','.','.','.','.'],
    ['.','9','1','.','.','.','.','.','3'],
    ['5','.','.','.','6','.','.','.','4'],
    ['.','.','.','8','.','3','.','.','5'],
    ['7','.','.','.','2','.','.','.','6'],
    ['.','.','.','.','.','.','2','.','.'],
    ['.','.','.','4','1','9','.','.','8'],
    ['.','.','.','.','8','.','.','7','9']]

    println(isValidSudoku(board))
end

